src/org/pentaho/reporting/platform/plugin/output/CachingPageableHTMLOutput.java

```diff
@@ -161,31 +178,87 @@ private IReportContent produceCacheablePages( final MasterReport report, final i
     final PageableHtmlOutputProcessor outputProcessor = (PageableHtmlOutputProcessor) proc.getOutputProcessor();
     outputProcessor.setFlowSelector( new DisplayAllFlowSelector() );
 
-    final Map<Integer, byte[]> pages = new HashMap<>();
+    final ExtendedConfiguration config = ClassicEngineBoot.getInstance().getExtendedConfig();
+    final boolean firstPageMode =
+      config.getBoolProperty( "org.pentaho.reporting.platform.plugin.output.FirstPageMode" );
+
+    //Async listener
+    final ReportProgressListener listener = ReportListenerThreadHolder.getListener();
```

Comment:
Listeners are informed in the order in which you add them. I'd add the cache-listener first, so that you can be sure that the content is written before you update the status. When you update the status, an eager user can try to fetch the content before it has been written. You can avoid that race condition by just having the correct order here.